DatabaseTarget - Added AllowDbNull for easier support for nullable pa…

snakefoot opened the pull request:
Instead of doing this:

`<parameter name="@exception" layout="${exception:format=tostring:whenEmpty=${db-null}}" parameterType="System.Object" 
/>`
`<parameter name="@correlationid" layout="${activityid:whenEmpty=${db-null}}" dbType="DbType.Guid" parameterType="System.Object" />`

Then one can do this:

`<parameter name="@exception" layout="${exception:format=tostring}" allowDbNull="true" />`
`<parameter name="@exception" layout="${activityid}" dbType="DbType.Guid" allowDbNull="true" />`

Trying to resolve #4070

304NotModified commented on src/NLog/Targets/DatabaseParameterInfo.cs:
```diff
@@ -134,6 +134,13 @@ public DatabaseParameterInfo(string parameterName, Layout parameterLayout)
+        /// </summary>
+        /// <docgen category='Parameter Options' order='8' />
+        [DefaultValue(false)]
+        public bool AllowDbNull { get; set; }
```
What about EmptyToDbNull or Nullable?

snakefoot replied:
Well in SQL language it is often called allow NULL. So that is why I like the **Allow**-part. But because in DotNet then it is not an actually null-reference but instead the artificial **DbNull**. This gives **AllowDbNull**

304NotModified replied:
Clear, but it does more than allow, it translates. I'm not sure how clear that is from the name (see also the description, it's using the term translate and not allow). But I also like to have feedback if others :)

snakefoot replied:
Have updated to source-code comment about having allow null configured for the database-column :)

snakefoot commented:
> And is this difficult to unit test?

I seldom write unit-tests when making prototypes. I have never used the DatabaseTarget for anything myself. I was actually hoping that you or @leonardder would chime in and acknowledge / reject the direction.

snakefoot commented:
@304NotModified Now added unit-test that puts some more concrete around the name `AllowDbNull` :)

leonardder commented:
Would this pr cover the case where

1. The database column allows nulls and is of a string type
2. An empty string (so not a null) is written to the target?

snakefoot commented:
NLog does not know the difference between null and string.Empty:


- AllowDbNull = false will change null into string.Empty when DbType is string (else default of the expected value type, ex 0 when integer)

- AllowDbNull = true will change null and string.Empty into DbNull


Sent from my Sony Xperia


---- Leonard de Ruijter wrote ----


Would this pr cover the case where

  1.  The database column allows nulls and is of a string type
  2.  An empty string (so not a null) is written to the target?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/NLog/NLog/pull/4076#issuecomment-671280736>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/ACXZ7HHSN3X6OMXSFHX2SZLR77EPTANCNFSM4PYUSZCQ>.

leonardder commented on tests/NLog.UnitTests/Targets/DatabaseTargetTests.cs:
```diff
@@ -607,7 +607,13 @@ public void LevelParameterTest(DbType? dbType, bool noRawValue, string expectedV
         [InlineData("${event-properties:almostAsIntProp}", DbType.AnsiString, " 124 ")]
-        public void GetParameterValueTest(string layout, DbType dbtype, object expected, bool convertToDecimal = false)
+        [InlineData("${event-properties:NullRawValue}", DbType.AnsiString, "")]
+        [InlineData("${event-properties:NullawValue}", DbType.Int32, 0)]
```
```suggestion
        [InlineData("${event-properties:NullRawValue}", DbType.Int32, 0)]
```

leonardder commented on tests/NLog.UnitTests/Targets/DatabaseTargetTests.cs:
```diff
@@ -607,7 +607,13 @@ public void LevelParameterTest(DbType? dbType, bool noRawValue, string expectedV
+        [InlineData("${event-properties:NullRawValue}", DbType.AnsiString, "")]
+        [InlineData("${event-properties:NullawValue}", DbType.Int32, 0)]
+        [InlineData("${event-properties:NullRawValue}", DbType.AnsiString, null, true)]
+        [InlineData("${event-properties:NullRawValue}", DbType.Int32, null, true)]
```
It would be nice if you could add tests for Guid as well:

```suggestion
        [InlineData("${event-properties:NullRawValue}", DbType.Int32, null, true)]
        [InlineData("${event-properties:NullRawValue}", DbType.Guid, null, true)]
```

leonardder commented on tests/NLog.UnitTests/Targets/DatabaseTargetTests.cs:
```diff
@@ -607,7 +607,13 @@ public void LevelParameterTest(DbType? dbType, bool noRawValue, string expectedV
+        [InlineData("${event-properties:NullRawValue}", DbType.AnsiString, null, true)]
+        [InlineData("${event-properties:NullRawValue}", DbType.Int32, null, true)]
+        [InlineData("", DbType.AnsiString, null, true)]
+        [InlineData("", DbType.Int32, null, true)]
```
```suggestion
        [InlineData("", DbType.Int32, null, true)]
        [InlineData("", DbType.Guid, null, true)]
```

snakefoot commented:
@leonardder Have fixed the spelling error, and added some testing of `DbType.Guid`

Introduced a small inconsistency, so if the `LogEventInfo.Properties`-dictionary contains `string.Empty` value, then it will respect the value and not convert into DbNull when using `AllowDbNull = true`.

304NotModified commented:
Merged! Great feature!
